### PR TITLE
it works on nodejs 16

### PR DESCRIPTION
**Pull request by koonopek**

### CODE REVIEW

**Comment by koonopek** [comment]
To do:
- [x] check if it works in snap
- [ ] make simple cli interface
- [ ] publish to npm

File: packages/wasm-bundler/index.mjs [hunk 1]
```diff
@@ -140,19 +101,16 @@ traverse(ast, {
...
-            // console.log({ node });
-            // console.log({ node: node.body.body });
             const newDeclaration = parse(wasmLoadFunc, { sourceType: 'module' });
-            // console.log({newDeclaration: newDeclaration.program.body[0].body.body});
             node.body.body = newDeclaration.program.body[0].body.body
-            // path.replaceWith(newDeclaration);
+        }
+        if (isFunctionDeclaration(node) && node.id.name === 'init') {
+            node.body.body[2] = parse("");
         }
```
Comment (piotr-roslaniec) [inline packages/wasm-bundler/index.mjs]: So the `init()` function is going to be removed, right? It contains a call to `getImports()`, can we do without them?

File: packages/wasm-bundler/wasm/index.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import { default as wasm0 } from './wasm0.js';
```
Comment (piotr-roslaniec) [inline packages/wasm-bundler/wasm/index.js]: This is an autogenerated file and it should be included in the `build` or `dist` directory on build time. Should probably add it to `.gitignore.` and/or change the output path in the `index.js` script.

File: packages/wasm-bundler/wasm/wasm0.js [hunk 3]
```diff
@@ -0,0 +1 @@
...IABuV/AB/1cAcAAHAC0BAQECAQIBAUgLMBUQAWUHAgYCAgEEIwEeG1sLOgkJARgEAQkBAwEFKwM8CCoYASA3AQEBBAgEAQMHCgIdAToBAQECBAgBCQEKAhoBAgI5AQQCBAICAwMBHgIDAQsCOQEEBQECBAEUAhYGAQE6AQECAQQIAQcDCgIeATsBAQEMAQkBKAEDATcBAQMFAwEEBwILAh0BOgECAQIBAwEFAgcCCwIcAjkCAQECBAgBCQEKAh0BSAEEAQIDAQEIAVEBAgcMCGIBAgkLBkoCGwEBAQEBNw4BBQECBQsBJAkBZgQBBgECAgIZAgQDEAQNAQICBgEPAQADAAMdAh4CHgJAAgEHCAECCwkBLQMBAXUCIgF2AwQCCQEGA9sCAgE6AQEHAQEBAQIIBgoCATAfMQQwBwEBBQEoCQwCIAQCAgEDOAEBAgMBAQM6CAICmAMBDQEHBAEGAQMCxkAAAcMhAAONAWAgAAZpAgAEAQogAlACAAEDAQQBGQIFAZcCGhINASYIGQsuAzABAgQCAicBQwYCAgICDAEIAS8BMwEBAwICBQIBASoCCAHuAQIBBAEAAQAQEBAAAgAB4gGVBQADAQIFBCgDBAGlAgAEAAKZCzEEewE2DykBAgIKAzEEAgIHAT0DJAUBCD4BDAI0CQoEAgFfAwIBAQIGAaABAwgVAjkCAQEBARYBDgcDBcMIAgMBARcBUQECBgEBAgEBAgEC6wECBAYCAQIbAlUIAgEBAmoBAQECBgEBZQMCBAEFAAkBAvUBCgIBAQQBkAQCAgQBIAooBgIECAEJBgIDLg0BAgAHAQYBAVIWAgcBAgECegYDAQECAQcBAUgCAwEBAQACAAU7BwABPwRRAQACAC4CFwABAQMEBQgIAgceBJQDADcEMggBDgEWBQEPAAcBEQIHAQIBBQAHAAE9BAAHbQcAYIDwAGNsb3N1cmUgaW52b2tlZCByZWN1cnNpdmVseSBvciBkZXN0cm95ZWQgYWxyZWFkeQBBuK7AAAsBAQB7CXByb2R1Y2VycwIIbGFuZ3VhZ2UBBFJ1c3QADHByb2Nlc3NlZC1ieQMFcnVzdGMdMS42My4wICg0YjkxYTZlYTcgMjAyMi0wOC0wOCkGd2FscnVzBjAuMTkuMAx3YXNtLWJpbmRnZW4SMC4yLjgyICg1OTg4M2VhY2Ep"; }
```
Comment (piotr-roslaniec) [inline packages/wasm-bundler/wasm/wasm0.js]: Same as above.

File: packages/wasm-bundler/wasm/index.js (hunk 2 above)
Comment (piotr-roslaniec) [reply packages/wasm-bundler/wasm/index.js]: Just to clarify, I mean it should end up in the `packages/rust-rng/pkg` directory.

File: packages/wasm-bundler/wasm/wasm0.js (hunk 3 above)
Comment (koonopek) [reply packages/wasm-bundler/wasm/wasm0.js]: sure

File: packages/wasm-bundler/wasm/index.js (hunk 2 above)
Comment (koonopek) [reply packages/wasm-bundler/wasm/index.js]: sure

File: packages/wasm-bundler/index.mjs (hunk 1 above)
Comment (koonopek) [reply packages/wasm-bundler/index.mjs]: Nope, only one if statment will be removed, which may call `fetch`.
This is fragile to changes in wasm-pack as everything  in this code base

File: packages/wasm-bundler/wasm/index.js (hunk 2 above)
Comment (koonopek) [reply packages/wasm-bundler/wasm/index.js]: I want to remove /wasm-bundler/ and pack everything in rust-rng, but first check if it works in snap

**Comment by koonopek** [comment]
When trying to load in snap:
` Build error: Unexpected token (179:46) while parsing /Users/michalkonopka/code/github.com/snap-rust-rng/rust-rng/pkg/rust_rng.js while parsing file: /Users/michalkonopka/code/github.com/snap-rust-rng/rust-rng/pkg/rust_rng.js`

Propably we still have to preprocess this file in some way to `umd` format or something

**Comment by piotr-roslaniec** [comment]
> Propably we still have to preprocess this file in some way to umd format or something

My best guess is that we need to do whatever `rollup.config.js` was doing previously

**Comment by koonopek** [comment]
Okey so it works in Snap! Now just to gather everything in nice tool and we are redy to ship!

File: packages/snap/src/index.ts [hunk 4]
```diff
@@ -1,31 +1,23 @@
 import { OnRpcRequestHandler } from '@metamask/snap-types';
-import {
-  InitOutput,
-  // add_random,
-  add_random_with_seed
-} from 'wasm-bundler';
-import { getRandomBytes } from './random';
 
-import { initializeWasm } from './wasm';
+import initWasm from '../../../rust-rng/dist/index';
```
Comment (piotr-roslaniec) [inline packages/snap/src/index.ts]: Ideally, we would be able to do just `import { add } from 'rust-rng'`

File: packages/snap/src/index.ts [hunk 5]
```diff
@@ -1,31 +1,23 @@
...
 
-let wasm: InitOutput;
 
 export const onRpcRequest: OnRpcRequestHandler = async ({
   origin,
   request,
 }) => {
   if (!wasm) {
-    wasm = await initializeWasm();
+    wasm = await initWasm();
```
Comment (piotr-roslaniec) [inline packages/snap/src/index.ts]: When using `smaldot` we don't have to manually init the WASM module like this. It tells me that we're not doing yet what they are doing. Maybe there is some room for further refinement.

**Comment by koonopek** [comment]
Actually they are doing this in polkadot.js, you have to call await
crypto.init() or something like that when using wasm

Pozdrawiam,
Michał Konopka

śr., 16 lis 2022, 08:46 użytkownik piotr-roslaniec ***@***.***>
napisał:

**Comment by piotr-roslaniec** [comment]
> Actually they are doing this in polkadot.js, you have to call await crypto.init() or something like that when using wasm

My bad, thanks!

Anyway, this is great work!

**Comment by koonopek** [comment]
But maybe we can do better, some day 😁

Pozdrawiam,
Michał Konopka

śr., 16 lis 2022, 08:52 użytkownik piotr-roslaniec ***@***.***>
napisał:

**Pull request closed without merging**